hrm-domain/hrm-service/src/sql/listToSql.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (C) 2021-2023 Technology Matters
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+
+/**
+ * Takes a table and a list of properties and returns a SQL snippet that
+ * can be used in a SELECT statement to select all of the properties from
+ * the table.
+ *
+ * @param {string} table - name of the table
+ * @param {string[]} fields - array of field names
+ * @returns {string} - SQL snippet
+ */
+export const fieldListToSql = (table: string, fields: readonly string[]): string =>
+  fields.map(prop => `"${table}"."${prop}"`).join(', ');
+
+/**
+ * Takes a table, the name of a jsonb field, and a list of properties in dot
+ * notation and returns a SQL snippet that can be used in a SELECT statement
+ * to return a JSON object with the properties from the jsonb field.
+ *
+ * @param {string} table - name of the table
+ * @param {string} field - name of the jsonb field
+ * @param {string[]} properties - array of property names in dot notation
+ * @returns {string} - SQL snippet
+ */
+export const objectNotationToBuildObjectSql = (
```

Comment:
In my mind, this pattern is a tiny initial step towards the rough permission division of list, view, edit. In the simple version of this pattern there are minimal "per field" permissions that get VERY complicated VERY quickly. The simplest form is just to leave out the data someone doesn't have access to from the list permission and reject their request for the view endpoint that would load data they don't have access to. I think other terminology is being used on the team, like "limited view" vs "full view", but the concept, in my mind, is similar.

This is the very tip of the iceberg related to a way of enforcing those permissions at the SQL level via abstraction and configuration in code.

Reply:
Whilst definitely think this approach is the way to go on the table field level, I just wonder whether the added complexity of generating Postgres JSON SQL queries vs doing client side transformations is worth the benefit. If we did those client transforms at the lowest layer it would still be abstracted from the perspective of the rest of the system I think?

I'm ok with either approach, I think both would work, just wanted to see why you thought filtering the JSON blobs in SQL was the better approach in this case